### ISSUE TITLE
Contributor Comment Financial Incentives - On Issue

### ISSUE THREAD

**Issue by pavlovcik** [ubiquity/ubiquibot#272]
### Financial Incentives for all Commenting Contributors

- The goal is to financially incentivize core contributors for their work beyond just the bounty hunter solving bounties.
- A core contributor is somebody who is added to the organization. They are identified at least as a `member`.
- We can calculate all comments posted by every core contributor in every issue and merged pull request. Once the bounty is solved (the pull request is merged) the bot should generate payment permits for every contributor all in the same permit comment.

### Default Formula

> Different elements in a contributor's comment corpus could have different price tags.
For example: plaintext=$0.1/word; link=$1; code block/image=$5;

_From contributor financial incentives discussion: https://github.com/ubiquity/bounty-bot/discussions/70_

Addendum 4 May 2023: 

- Formatting with lists, e.g. each bullet point or numbered point can be credited $0.50 and then the words after with normal $0.10 calculation.
- Double the credit in the original post (e.g. the bounty specification should receive double credits like `$0.2/word` etc. to try and compensate for the time to think of the bounty in the first place.)

Addendum 17 May 2023

- Read repository `BASE_RATE` from the `ubiquibot-config.yml` and multiply the output. For example, our default `BASE_RATE` IS `1000` and the output of the function should return `1` to multiply against. In our Ubiquity Dollar repository, we need more attention on it so I increased the `BASE_RATE` to `2000`. This means that bounties and comment incentives all should double in order to incentivize more help and attention on that repository. 
- Added pull request page calculations specification #311

### Specification 

- We should use the above prices as default, but also allow each to be configurable via the config yml file. 
  - We should also be able to fully disable this feature (set values all to 0?) in case organizations only want to pay bounty hunters.
- In the payment permit comment, the bot should list permits for every contributor. 
  - The bot should total all comments from all of the following parties (exceptions listed below) to reward their efforts for their part in completing the bounty: 
    - In the pull request review: 
      - All of the reviewers.
    - In the issue: 
      - The issue commenters, especially the issue creator. 

### Exceptions

- Bounty Hunter - Their only payment should be for the price of the bounty! They should not receive additional compensation for asking questions etc. (this is too easily exploitable.)
- Any bot commands - if somebody is registering their wallet address or calling the help function, they shouldn't get paid for it!
- Quoted text - Should not be credited because 1. it's easy to abuse with layered quoted text (GitHub conversation experience kind of encourages nested quotes) 2. it is, by definition, not any new information that is provided by the bounty hunter, so they shouldn't get credit for it.

#### Configurations
Technically we should be able to find an open source list of every recognized GitHub flavored markdown entity e.g. headers, bolded words, links, images, code blocks etc. and then be able to override and specify the price. For now, we should at least support overrides for the following:
- word (the baseline/default)
- link
- codeblock
- image
- list item (numbered)
- list item (disc)
- quote

**Comment by pavlovcik** [comment]
I think it would be interesting to financially incentivize higher quality posts. This is not apart of the spec, but just brainstorming down here:

1. Formatting like with lists, perhaps each bullet point or numbered point can be credited $0.50 and then the words after with normal $0.10 calculation.
2. Not recognized words could be considered as typos and not credited for.

**Comment by pavlovcik** [comment]
For researchers to get paid, it could be really interesting in the future to see discussions converted into issues, and then tallying all of the contributions in the discussion as part of the "pre-bounty-spec" permit totals. 

By definition that discussion with all of the contributors' work directly led to the bounty specification, which should be rewarded!

**Comment by pavlovcik** [comment]
Last remark for now, perhaps for a future bounty. 

Non core team should be able to file issues and eventually be rewarded. 

1. They open a new issue
2. It is reviewed by the team and they eventually fund it (make it into an official bounty)
3. Bot asks for community members address 
4. Bounty eventually gets solved 
5. Everybody including the public customer/user is rewarded. 

This could make product development go into hyper speed if our customers could vote on features and be financially incentivized to let us know. Would be epic to sort bounty priority order automatically based on customer demand (like they can vote on certain issues to prioritize)

Perhaps that could be a novel utility of our governance tokens, but more realistically each project's unique governance tokens.

**Comment by pavlovcik** [comment]
@0xcodercrane I slightly updated the specification to include credit for lists as well. Check for 

> Addendum 4 May 2023:

**Comment by 0xcodercrane** [comment]
> Releasing the bounty back to dev pool because the allocated duration already ended!

Apologize ser. but I want to complete it myself lol

**Comment by rndquu** [comment]
Getting content from github markdown is tricky. I don't know what solution @0xcodercrane will choose in the end but the best thing I've found so far is https://www.npmjs.com/package/remark-parse. So it would be really great if we could convert any github markdown to AST, smth similar to:
```
{
  headings: [
   { value: "Heading 1" },
   { value: "Heading 2" }
 ],
 lists: [
   { items: ["item 1", "item 2"] },
   { items: ["item 3", "item 4"] }
 ]
}
``` 

So in the end we could assign price for each parsed AST element

**Comment by rndquu** [comment]
@0xcodercrane here is a more practical example

Check this [comment](https://github.com/ubiquity/ubiquibot/issues/272#issuecomment-1522573884)

This code transforms the above comment to [mdast](https://github.com/syntax-tree/mdast):
```
import { fromMarkdown } from 'mdast-util-from-markdown';
import { gfmFromMarkdown } from 'mdast-util-gfm';
import { gfm } from 'micromark-extension-gfm';
import fs from 'node:fs/promises';

async function main() {
    const comment = await fs.readFile('comment.md');

    const tree = fromMarkdown(comment, {
        extensions: [gfm()],
        mdastExtensions: [gfmFromMarkdown()]
    });

    console.log(JSON.stringify(tree));
}

main();
```

This is how the result markdown abstract syntax tree looks like: https://gist.github.com/rndquu/76161e70421bb7e687487dab8f353db2

There are different types of parsed content:
- paragraph
- text
- link
- etc...

So we could assign a price for a particular node type (text, link, etc...). 

For example for `text` or `paragraph` types we could assign 0.01$ and then count the number of symbols. So this would be basically "price per symbol".

For the `link` type we could assign 0.5$ which would mean 0.5$ for a whole URL.

We can keep it simple and start from "text", "paragraph" and "link" types. I guess this task implies building a "framework" for such calculations rather than covering all possible node content types. We can add pricing for different node types later.

**Comment by pavlovcik** [comment]
> I guess this task implies building a "framework" for such calculations rather than covering all possible node content types. We can add pricing for different node types later.

I agree with this, but I also would really like to cover the elements in the specification I produced.

**Comment by 0xcodercrane** [comment]
> > I guess this task implies building a "framework" for such calculations rather than covering all possible node content types. We can add pricing for different node types later.
> 
> I agree with this, but I also would really like to cover the elements in the specification I produced.

The items you produced needs to be matched to what the mdast supports in the beginning. but I guess it has most of things we want.

**Comment by Steveantor** [comment]
> Releasing the bounty back to dev pool because the allocated duration already ended!

there should be an assign locking function

**Comment by pavlovcik** [comment]
It's supposed to be able to monitor progress on the linked pull requests instead.

**Comment by 0xcodercrane** [comment]
> Do you have any updates @0xcodercrane? If you would like to release the bounty back to the DevPool, please comment `/unassign`

Lol yes you need to track the linked PR as well

**Comment by Steveantor** [comment]
When next bot release?

**Comment by pavlovcik** [comment]
> When next bot release? 

Wen? Soon™️ 

Track the release progress here #292

**Comment by Steveantor** [comment]
> > When next bot release?
> 
> Wen? Soontm
> 
> Track the release progress here #292

Fix the CI and merge?
I don't see any progress there

**Comment by pavlovcik** [comment]
> > > When next bot release?
> 
> > 
> 
> > Wen? Soontm
> 
> > 
> 
> > Track the release progress here #292
> 
> 
> 
> Fix the CI and merge?
> 
> I don't see any progress there

😂 sick burn 

@0xcodercrane

**Comment by pavlovcik** [comment]
I want to also test this feature. Where's the best place I can observe an example?

**Comment by 0xcodercrane** [comment]
> I want to also test this feature. Where's the best place I can observe an example?

Sure. there is a [QA](https://github.com/ubiquibot/staging/issues/14) but I will let you know once I finish the QA myself because there could be a couple of refactors. until everything is clear, leave the claim url here.

**Comment by 0xcodercrane** [comment]
> 



> ### [ **[ CLAIM 600 DAI ]** ](https://pay.ubq.fi?claim=eyJwZXJtaXQiOnsicGVybWl0dGVkIjp7InRva2VuIjoiMHg2QjE3NTQ3NEU4OTA5NEM0NERhOThiOTU0RWVkZUFDNDk1MjcxZDBGIiwiYW1vdW50IjoiNjAwMDAwMDAwMDAwMDAwMDAwMDAwIn0sIm5vbmNlIjoiMjI5Nzk4NTU4ODU3MDA1MTk5NjM4ODM0OTE0NDIzMTQ0MDU4MzIzOTE2NDA3MTM1MDUyMTQ5NjEzNzI4MjMxNzUyOTAwOTE1NjY3NDEiLCJkZWFkbGluZSI6IjExNTc5MjA4OTIzNzMxNjE5NTQyMzU3MDk4NTAwODY4NzkwNzg1MzI2OTk4NDY2NTY0MDU2NDAzOTQ1NzU4NDAwNzkxMzEyOTYzOTkzNSJ9LCJ0cmFuc2ZlckRldGFpbHMiOnsidG8iOiIweDAwODY4QkIzQkEyQjM2MzE2YzJmYzQyRTRhRkI2RDQyNDZiNzdFNDYiLCJyZXF1ZXN0ZWRBbW91bnQiOiI2MDAwMDAwMDAwMDAwMDAwMDAwMDAifSwib3duZXIiOiIweGY4N2NhNDU4M0M3OTIyMTJlNTI3MjBkMTI3RTdFMEEzOEI4MThhRDEiLCJzaWduYXR1cmUiOiIweDg4NjY5Y2MwY2E1NDU5MTY2YzgyOGJhMmQ3NWNhMTI4NTZhZjBlNTI1MjczZjYzOGExYWZiOTliZmE3ZGIyYzY1OWJiYjU5MTNlM2FhNWVmMWNiMWNkY2QwNDk4Y2VhZTJiYmFhYzk0ZTZiM2ZjNmMxZmRkNzNkNTEwNmZlODcxMWIifQ==)
> `0x00868BB...46b77E46`

Keep it here until the QA passes

**Comment by Ameralameri** [comment]
> ### الأوامر المتوفرة
> ```
> - /assign: Assign the origin sender to the issue automatically.
> - /unassign: Unassign the origin sender from the issue automatically.
> - /help: List all available commands.
> - /payout: Disable automatic payment for the issue.
> - /multiplier: Set bounty multiplier (for treasury)
> - /allow: Set access control. (Admin Only)
> - /wallet: <WALLET_ADDRESS | ENS_NAME>: Register the hunter's wallet address. 
>   ex1:wallet 0x20fe9f5896a3021ca6ea78005a15ae954dc134e6
>   ex2: /wallet vitalik.eth
> ```
> 
> @Ameralameri

**Comment by 0xcodercrane** [comment]
Status: The basic flow is done, but in the QA step, it was not working properly.
Claim status: not claimed any funds.

**Comment by wannacfuture** [comment]
assignComment

**Comment by pavlovcik** [comment]
> @wannacfuture please be sure to review this conversation and implement any necessary fixes. Unless this is closed as completed, its payment of **600.0 DAI** will be deducted from your next bounty.

lol @0xcodercrane you just rugged @wannacfuture btw

**Comment by pavlovcik** [comment]
@wannacfuture there is a bug with the bot, but I want you to comment so that I can assign you and then mark this as completed for auditing purposes. 

Please address the above steps when you see this thanks.

**Comment by wannacfuture** [comment]
Let me take this.
